### PR TITLE
[Yaml] respect inline level when dumping objects as maps

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22392
| License       | MIT
| Doc PR        |

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by stof** [comment]
Please also add a test with an empty ArrayObject or stdClass, to ensure it keeps being dumped as `{}` and not `[]`.

**Comment by xabbuh** [comment]
done

**Comment by goetas** [comment]
Hi, I'm not sure is the best approach to solve the issue. 
The following test will fail:

```php
public function testDumpingArrayObjectInstancesWithNumericKeysRespectsInlineLevel()
{
    $deep = new \ArrayObject(array('d', 'e'));
    $inner = new \ArrayObject(array('b', 'c', $deep));
    $outer = new \ArrayObject(array( 'a', $inner));

    $yaml = $this->dumper->dump($outer, 2, 0, Yaml::DUMP_OBJECT_AS_MAP);
    $expected = <<<YAML
0: a
1:
    0: b
    1: c
    2: { 0: d, 1: e }

YAML;
    $this->assertEquals($expected, $yaml);
}
``` 

and the strategy  (casting array object ot array) will make really hard to solve this case

**Comment by goetas** [comment]
I will also try to find a solution

**Comment by xabbuh** [comment]
Status: Needs work

**Comment by goetas** [comment]
Here is my version of the fix https://github.com/symfony/symfony/pull/22419

**Comment by goetas** [comment]
Question: is https://github.com/symfony/symfony/pull/21471 considered a new feature? IMO should be a bugfix and included in the 3.2 branch.

Empty arrays should be serialized as `[ ]`, and empty arrayobject instances should be serialized as `{ }`, the same behavior is used with `json_encode`

**Comment by stof** [comment]
> Question: why #21471 is considered as a new feature?

This relies on adding a new constant for the configuration flag => this is a new feature for the component

**Comment by goetas** [comment]
@xabbuh  did you check my work at https://github.com/symfony/symfony/issues/22419

**Comment by goetas** [comment]
@stof  i see, thanks

**Comment by goetas** [comment]
@xabbuh if this is the candidate PR to be merged, do you mind adding:

```php
    public function testDumpingArrayObjectInstancesWithNumericKeysRespectsInlineLevel()
    {
        $deep = new \ArrayObject(array('d', 'e'));
        $inner = new \ArrayObject(array('b', 'c', $deep));
        $outer = new \ArrayObject(array('a', $inner));
        $yaml = $this->dumper->dump($outer, 2, 0, Yaml::DUMP_OBJECT_AS_MAP);
        $expected = <<<YAML
0: a
1:
    0: b
    1: c
    2: { 0: d, 1: e }
YAML;
        $this->assertEquals($expected, $yaml);
    }
```
?

File: src/Symfony/Component/Yaml/Dumper.php [hunk 1]
```diff
@@ -98,11 +103,17 @@ public function dump($input, $inline = 0, $indent = 0, $flags = 0)
                     continue;
                 }
 
-                $willBeInlined = $inline - 1 <= 0 || !is_array($value) || empty($value);
+                $dumpObjectAsInlineMap = true;
+
+                if (Yaml::DUMP_OBJECT_AS_MAP & $flags && ($value instanceof \ArrayObject || $value instanceof \stdClass)) {
```
Comment (goetas) [inline src/Symfony/Component/Yaml/Dumper.php]: Should this code be put in a private static method? is identical to what we have on line 86

**Comment by goetas** [comment]
Did not know `stdClass` looks to be "iterable" even if `is_iterable` says is not https://3v4l.org/bY9IC

**Comment by stof** [comment]
@goetas it does not pass the `iterable` typehint either. but iterating over a non-iterable object iterates over public properties in PHP.

**Comment by goetas** [comment]
@stof Is so beautiful to learn things about PHP every day after so many years :relaxed:

**Comment by stof** [comment]
See https://3v4l.org/tu1Z0 for the proof btw

**Comment by goetas** [comment]
Do this PR need still work? 

For me is :+1:

**Comment by fabpot** [comment]
Thank you @xabbuh.
